Add primitive String type

mirryi opened the pull request:
This PR adds a primitive string type, string literals, and builtin operations on those strings:

- String literals 
  - [x] OCaml escape sequences (`"abc\t\n")
    - [x] Strings with invalid escapes are indet values 
  - [x] Evaluation result syntax highlighting
    - [x] Invalid escape sequences

- String operations
  - [x] Binary concatenation operator (`s1 ^ s2`)
    - [x] Doesn't work on indet values (maybe it should)
  - [x] Subscript operator (`s1[n1:n2]`)
    - [x] Correct error handling and indicators for out-of-bound indices
    - [x] Also doesn't work on indet values (maybe it should)
  - Builtin functions for reading strings
    - [x] `length : String -> Int` returns length of a string
    - [x] `{int,float,bool}_of_string : String -> {Int,Float,Bool}` parses a string as an integer/float/bool (can fail)
    - [x] `string_of_{int,float,bool} : {Int,Float,Bool} -> String` converts an integer/float/bool to a string
  
- String pattern matching
  - [x] String literal patterns

- Parser integration
  - [x] String literal expressions and patterns
  - [x] Binary concatenation operator
  - [x] Mix-fix subscript operation

Leaving off for now:

- Code syntax highlighting

mirryi committed:
Add String variant to HTyp.t and UHTyp.operand

mirryi committed:
Progress on String Action semantics

mirryi committed:
Fix UHExp.is_complete_operand for StringLit

mirryi committed:
Reorder String and Hole variant in HTyp

mirryi committed:
Fix ZExp helpers on StringLit

mirryi committed:
Fix typo in comment

mirryi committed:
Implement ZExp.ana_perform_operand

mirryi committed:
Handle strings in web doc

mirryi committed:
Add SQuote to ActionPanel

mirryi committed:
Handle SQuote in undo history

mirryi committed:
Add key combo to insert string literal

mirryi committed:
Fix cursor position after inserting string lit exp and pat

mirryi committed:
Disallow SQuote inside StringLit exp or pat

mirryi committed:
Fix UndoHistoryCore.cursor_term_len for StringLit

mirryi committed:
Fix basic ops for StringLit ZExp and ZPat

mirryi committed:
Add todo in UHDoc_common.re

mirryi committed:
Implement StringLitLexer

mirryi committed:
Add support for decimal, hex, octal string escape sequences

mirryi committed:
Fix UndoHistoryCore.cursor_term_len for StringLit Pat

mirryi committed:
Fix string literal display

mirryi committed:
Fix keybinding for SQuote construction

mirryi committed:
Fix input of key combo keys

mirryi committed:
Change UHExp.StringLit inner to EscapedString

mirryi committed:
Remove EscapedString.concat

mirryi committed:
Implement string lit elaboration

mirryi committed:
Skip error handling for string lit elaboration

mirryi committed:
Add concatenation operator and implement some statics

mirryi committed:
Rename Caret op to SCaret

mirryi committed:
Add SCaret to DHExp and implement elaboration

mirryi committed:
Implement string concat evaluation

mirryi committed:
Fix hazelweb display components for SCaret

mirryi committed:
Annotate new InvalidInput codes

mirryi committed:
Add keybinds for caret action

mirryi committed:
Add Subscript to UHExp and ZExp, implement movements

mirryi committed:
Implement CursorInfo_Exp for subscript

mirryi committed:
Implement CursorPath_Exp for subscript

mirryi committed:
Implement SubscriptZE syn action

mirryi committed:
Finish actions for subscript

mirryi committed:
Implement Subscript type statics

mirryi committed:
Implement cursor_term_len for subscript

mirryi committed:
Add subscript elaboration

mirryi committed:
Implement subscript evaluation

mirryi committed:
Fix gui errors for subscript

mirryi committed:
Add SSubscript construct action

mirryi committed:
Fix last hazelweb things for subscript

mirryi committed:
Fix subscript string padding

mirryi committed:
Disallow subscripting empty strings

mirryi committed:
Fix constructing nil list in hole

mirryi committed:
Fix some of subscript action cursor placement

mirryi committed:
Fix cursor placement in subscript

mirryi committed:
Don't delete the whole term when deleting subscript delim

mirryi committed:
Make subscript display more meaningful

mirryi committed:
Last, for ana

mirryi committed:
Go through added TODOs and remove some

mirryi committed:
Add Builtins.re with some basic builtin functions

mirryi committed:
Implement evaluation for builtins

mirryi committed:
Replace subst_var_list with List.map

mirryi committed:
Implement ui components of ApBuiltin

mirryi committed:
Set empty Context to builtins

mirryi committed:
Substitute builtin function name with wrapping lam

mirryi committed:
Fix builtin function substitution

mirryi committed:
Add todo notice for builtin substitution

cyrus- committed:
fix parenthesization in add_decimal_code

mirryi committed:
Rename Contexts.empty to Contexts.initial

mirryi committed:
Move Evaluator.result to own module

mirryi committed:
Move builtin implementations to Builtins.re

mirryi committed:
Use Ap case for ApBuiltin in DHDoc_Exp.mk

mirryi committed:
Merge branch 'dev' into builtin-fns

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Add string lit suggestion to empty hole in strategy guide

mirryi committed:
Reorganize Builtins to require registering new builtin fn in one place

mirryi committed:
Allow shadowing of builtins by inserting at beginning after elaboration

mirryi committed:
Merge branch 'dev' into builtin-fns

mirryi committed:
Merge branch 'builtin-fns' into strings4

mirryi committed:
Remove Evaluator comment with error codes

mirryi committed:
Fix CursorInfo_Exp.ana_cursor_info_zoperand for Subscript

mirryi committed:
Remove FIXME from UnescapedString

mirryi committed:
Add UnescapedString.from_string_unchecked

mirryi committed:
Add string_of_{int, bool, float} builtins

mirryi committed:
Add int_of_string builtin

mirryi committed:
Rewrite builtins using new mk_one_arg

mirryi committed:
Add {float, bool}_of_string builtins

mirryi committed:
Add string length builtin

mirryi committed:
Make DHExp.StringLit carry escape sequence errors

mirryi committed:
Highlight invalid escape sequences in string literals

This doesn't work properly yet because of some indexing shenanigans.
Probably going to change up how strings are represented overall to make
error propogation easier.

mirryi committed:
Add mk_one_arg helper for creating one arg builtins

mirryi committed:
Automatically evaluate d1 in mk_one_arg

mirryi committed:
Fix string lexing for error positions

mirryi committed:
Merge branch 'builtin-fns' into strings4

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Allow any character to be inserted into string literal

mirryi committed:
Fix inserting special characters in string literal and comment line

mirryi committed:
Remove string lit check in KeyComboAction.get_model_action_from_kc

mirryi commented:
There are still some issues that I observed from the Hazel assistant's popup with regards to type analysis, but otherwise I think everything is pretty much here?

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Add string built-ins back in

mirryi committed:
Rename built-in string_length to string

mirryi committed:
Merge branch 'dev' into strings4

cyrus- commented:
@Dophin2009 sorry for the delay in reviewing this. can you get it up to date with dev again?

mirryi committed:
Merge branch 'dev' into strings4

mirryi commented:
Feeling especially not confident about this code now after a few months...

cyrus- commented:
Do you want to give it another pass or should I review it soon?

mirryi commented:
Probably a good idea. I'll give it another pass today/tomorrow.

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Add documentation for StringLitLexer

mirryi committed:
Add documentation for UnescapedString

mirryi committed:
Clean up CursorInfo_Exp and CursorInfo_Pat

mirryi committed:
fixup! Add documentation for UnescapedString

mirryi committed:
Clean up various modules related to subscript

mirryi committed:
Clean up StringLit cursor position things

mirryi committed:
Clean up ZExp and ZPat

mirryi committed:
Clean up dynamics for strings

mirryi committed:
Add Invalid*OfString error kinds to DHAnnot

mirryi committed:
Clean up keyboard inputs for string

mirryi committed:
Fix builtin functions' types

mirryi committed:
Clean up doc/layout things

mirryi commented:
@cyrus- Should be good for a review now, I think!

cyrus- commented:
In the previous version of the strings branch, we had syntax highlighting for escape characters, including red for invalid escape characters. That seems to be missing here.
![Screenshot from 2022-06-15 16-50-33](https://user-images.githubusercontent.com/280638/173924987-71c519e9-aa61-4395-be72-b4e350961d7e.png)

mirryi committed:
WIP Add highlighting for string literal errors and escape seqs

mirryi committed:
fixup! WIP Add highlighting for string literal errors and escape seqs

mirryi commented:
I still have next to no real clue how the layout/doc system works, so I'm having a problem with implementing the syntax highlighting:

![image](https://user-images.githubusercontent.com/38477923/173969881-bd01d733-f868-4635-9bb9-21dd68e1b971.png)

If you were to hit the right arrow key when the cursor is right after `abc` and before `\`, it'd jump back to the beginning of the string, right after the quotation mark, instead of to the right of `\`. I think it's because the `abc` and `\` segments are two different `Doc.t` or something?

I modeled [the doc code](https://github.com/hazelgrove/hazel/blob/f07e4dd1ee1389d899e3a6ff7f969e2fa6afbd60/src/hazelweb/layout/UHDoc_common.re#L340) off of [the code in strings3](https://github.com/hazelgrove/hazel/blob/255516cac90a00bcd7d2189599c8edd2591ecf9c/src/hazelweb/layout/UHDoc_common.re#L176), but didn't want to repeat the parsing logic (which is already in `StringLitLexer`). Following strings3, I modified the [`Text` variant of `UHAnnot.token_shape`](https://github.com/hazelgrove/hazel/blob/f07e4dd1ee1389d899e3a6ff7f969e2fa6afbd60/src/hazelweb/layout/UHAnnot.re#L10) by adding a start index field, but then I'm not really sure what to do with it (e.g. [in `UHMeasuredLayout.re`](https://github.com/hazelgrove/hazel/blob/f07e4dd1ee1389d899e3a6ff7f969e2fa6afbd60/src/hazelweb/layout/UHMeasuredLayout.re#L23)).

mirryi commented:
Also, I fixed the escape sequence error highlighting for the evaluation code box.

cyrus- commented:
Hm, not sure what the problem is. can you point me to the relevant code?

cyrus- commented:
You can look at Mengjing's strings branch, I believe she had it working before...

mirryi commented:
Yeah, it's the code I linked in the above comment (`UHDoc_common`). I've been going off the strings3 branch, which seemed to duplicate a bunch of escape sequence parsing logic for syntax highlighting purposes? I made it so that lexer already produces the information necessary for that. 

If you were to make `inner` in `UHDoc_common.mk_StringLit` to just `mk_text(s)`, it would have no syntax highlighting but work fine. I followed strings3 by changing `UHAnnot.token_shape` like I said above, but the layout/doc code has changed a lot since then, so I'm not sure how to use it to make movements work correctly.

cyrus- commented:
Hm, I'm not super familiar with that code either. @disconcision can you take a quick look?

disconcision commented:
I took a quick look but don't really have an idea of what's going on... I have basic understanding of the layout system but I've never touched measured layout. I don't immediately understand the way strings are constructed here, I'd have to dig into the code in-depth before I could start asking the right questions.

From the nature of the bug (looping on first segment as opposed to getting stuck or skipping out of the string) I would tend to guess that there is something like two indices to represent the position in a string, the segment and the position within the segment, and the latter is being correctly reset at the end of a segment but the former is erroneously not being incremented. But I had trouble finding where in the code the relevant logic is done.

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Kind of fix invalid escape highlighting movement issues

mirryi committed:
Clean up DHDoc_common.mk_StringLit

mirryi committed:
Add type aliases for StringLitLexer items in UnescapedString

mirryi committed:
Add length, olength properties to StringLitLexer.seq

mirryi committed:
Rename StringLitLexer.{seq,error} to StringLitLexer.{valid_seq,invalid_seq}

mirryi committed:
Rename StringLitLexer to UnescapedStringParser

mirryi committed:
Extract UnescapedStringParser state to separate module

mirryi committed:
Refactor string literal parser state

mirryi committed:
Clean up UnescapedStringParserState public API

mirryi committed:
Clean up string parsing

cyrus- commented:
@mirryi mark as ready for review when its good to go again

mirryi committed:
Rename instances of errors to iseqs

nj-wilson committed:
Hazeltext - Backend (#598)

* Hazeltext - Backend

Adds Hazeltext for parsing strings into hazel ASTs.
Reworks a few interals to accommodate this, mainly splitting
out memoization, and moving hazelweb/layout to hazelcore/layout.

Rewrites Examples to be strings ran through the parser.

* ExternalLineBreak -> ExpLineBreak

* Lexer: Handle carriage return

Handles \r\n on Windows

Co-authored-by: Cyrus Omar <cyology@gmail.com>

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Add string literal parsing

mirryi committed:
Add string type parsing

mirryi committed:
fixup! Add string literal parsing

mirryi committed:
Add string concatenation parsing

mirryi committed:
Remove code syntax highlighting

mirryi committed:
WIP Add subscript parsing

mirryi committed:
Use stdlib result for UnescapedString.subscript return

mirryi committed:
Merge remote-tracking branch 'origin/dev' into strings4

mirryi committed:
Rename MetaVarGen in prep for merge

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Remove IDGen.rei and add back MetaVarGen.re/rei

mirryi committed:
Fix final merge issues

mirryi committed:
Fix subscript parsing

cyrus- commented:
A few things that we should fix when strings are reimplemented via tylr, but that we can pass on for now:
* highlighting of escape sequences and bad escape sequences
* be able to press enter within a string literal
* think about how we want to handle closing quotation marks vs. internal quotation marks (currently, closing quotation mark is automatic so `"` key inserts the quotation mark character
* undo history shows the entire string literal in the message even if it is very long, probably want to elide after a certain length and show only the part that changed
* ~~term highlighting on subscript is incorrect, should show start and end indices as open children~~

However, I'm okay with merging it pending a code review so we have baseline behavior specification for the tylr strings effort in dev.

cyrus- commented:
thing we should fix now: the `length` built in function doesn't return an integer or error when given a string with an invalid escape sequence in it

![Screenshot from 2022-07-14 00-06-47](https://user-images.githubusercontent.com/280638/178896281-e042bb0e-029c-4894-a95c-59dcca3b674c.png)

mirryi committed:
Fix length builtin for indet arguments

mirryi committed:
Add comments for string builtin implementations

mirryi committed:
Rename DHAnnot.InvalidStringEscape to DHAnnot.InvalidStringSeq

mirryi committed:
Merge remote-tracking branch 'origin/dev' into strings4

mirryi committed:
Remove various TODOs

cyrus- commented on src/hazelcore/UHExp.rei:
```diff
@@ -17,10 +17,12 @@ and operand =
   | IntLit(ErrStatus.t, string)
   | FloatLit(ErrStatus.t, string)
   | BoolLit(ErrStatus.t, bool)
+  | StringLit(ErrStatus.t, string)
```
string argument here should really be of type UnescapedString.t, right?

mirryi replied:
No, since the strings in UHExp are unparsed. They're parsed during elaboration. Maybe "unescaped" isn't the clearest naming, though.

cyrus- replied:
Oh I see, the name makes sense but maybe we should also define EscapedString.t so its clear which is which

mirryi committed:
Add .ocamlformat

cyrus- commented on src/hazelcore/UHExp.re:
```diff
@@ -19,10 +19,12 @@ and operand =
   | Fun(ErrStatus.t, UHPat.t, t)
   | Inj(ErrStatus.t, InjSide.t, t)
   | Case(CaseErrStatus.t, t, rules)
+  | Subscript(ErrStatus.t, t, t, t)
```
It might make more sense to have the subscript target be of type operand rather than t, since we don't allow it to be an opseq or any other kind of block expression anyway.

mirryi replied:
Not sure how to resolve one part of Action_Exp with this change, since `ana_perform` returns a `ZExp.t`, which I'm not sure how to fit back into `ZExp.SubscriptZE1` (the target of which I changed to `ZExp.zoperand`): [this part of the function](https://github.com/hazelgrove/hazel/blob/1d76b18930e2e17026041db617488df91e6975c9/src/hazelcore/Action_Exp.re#L2059-L2098), and a similiar part down below.

cyrus- commented on src/hazelcore/ZExp.re:
```diff
@@ -626,7 +697,10 @@ and move_cursor_left_zoperand =
   | CursorE(OnOp(_), _) => None
-  | CursorE(OnText(j), e) => Some(CursorE(OnText(j - 1), e))
+  | CursorE(OnText(j), e) => {
+      print_endline(string_of_int(j));
```
remove

cyrus- commented on src/hazelcore/ZExp.re:
```diff
@@ -626,7 +697,10 @@ and move_cursor_left_zoperand =
-  | CursorE(OnText(j), e) => Some(CursorE(OnText(j - 1), e))
+  | CursorE(OnText(j), e) => {
+      print_endline(string_of_int(j));
+      Some(CursorE(OnText(j - 1), e));
```
doesn't this break when j = 0?

mirryi replied:
Hmm, looks like [this has always been here](https://github.com/hazelgrove/hazel/blob/dev/src/hazelcore/ZExp.re#L629=). I guess we should add a guard to do the above when `j > 0` and return `None` otherwise?

cyrus- replied:
Yeah, though analogous issue happens for moving right and there you'd have to actually look at the rightmost valid position (which is returned from the valid cursor positions function)

cyrus- commented on src/hazelcore/ZExp.re:
```diff
@@ -805,7 +904,10 @@ and move_cursor_right_zoperand =
   | CursorE(OnOp(_), _) => None
-  | CursorE(OnText(j), e) => Some(CursorE(OnText(j + 1), e))
+  | CursorE(OnText(j), e) => {
+      print_endline(string_of_int(j));
```
remove

cyrus- commented on src/hazelcore/ZExp.re:
```diff
@@ -805,7 +904,10 @@ and move_cursor_right_zoperand =
-  | CursorE(OnText(j), e) => Some(CursorE(OnText(j + 1), e))
+  | CursorE(OnText(j), e) => {
+      print_endline(string_of_int(j));
+      Some(CursorE(OnText(j + 1), e));
```
what about if you hit the right side?

cyrus- commented on src/hazelcore/ZPat.re:
```diff
@@ -330,7 +337,8 @@ and move_cursor_right_zoperand =
       InvalidText(_, _) | Var(_, _, _) | BoolLit(_, _) | IntLit(_, _) |
-      FloatLit(_, _),
+      FloatLit(_, _) |
+      StringLit(_, _),
```
String literals have delimiters...

cyrus- commented on src/hazelcore/ZPat.re:
```diff
@@ -269,7 +275,8 @@ and move_cursor_left_zoperand =
+      FloatLit(_, _) |
+      StringLit(_, _),
     ) =>
     // invalid cursor position
```
String literals have delimiters...

cyrus- reviewed (changes requested):
functionality seems good now, partial review of syntax

mirryi committed:
Remove debug prints and prevent moving off left side of OnText

mirryi committed:
Fix move_cursor_right_operand for OnText positions

mirryi committed:
Fix StringLit movement

mirryi committed:
Make UHExp/UHPat.StringLit internal string an EscapedString.t

mirryi committed:
Check for bad StringLit OnDelim indices when constructing quote

mirryi committed:
Fix ZPat.StringLit movement

mirryi committed:
fixup! Fix ZPat.StringLit movement

mirryi committed:
Fix(?) subscript index highlighting

mirryi committed:
Merge branch 'dev' into strings4

mirryi committed:
Fix(?) subscript operand highlighting

mirryi committed:
Merge remote-tracking branch 'origin/dev' into strings4

mirryi committed:
Add back string tests to parsing tests

mirryi committed:
fixup! Add back string tests to parsing tests

mirryi committed:
Merge branch 'dev' into strings4

cyrus- commented:
Archiving this Hazel 2 PR, but see #786 -- this PR had many more features than the very basic string literals we have in Hazel 3 right now, so it should serve as a reference for how we eventually want to realize strings. Thank you all for the work on this!
